### DISCUSSION TITLE
Migration from Cldr to Intl API

### DISCUSSION THREAD

**Discussion by goldim** [qooxdoo/qooxdoo#10663] [General]
I try to implement `qx.locale` classes with [Intl API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) and get rid of `node-cldr` and `cldr-data`. Now I've encountered the next issues:

- [x] `getPercentFormat` method of `qx.locale.Number` - just still not done. There are no tests for this method.
- [x]  I have only one failed test currently with **de_DE** locale. Intl API has 3 types of weekday: narrow, short, long. Narrow - first letter of weekday, short - usually 2 or 3 letter words, and long - full name. In Cldr there abbreviated, wide and abbreviated. Cldr has another short option but it is not used. I tested with other like french, russian, english and they are fine.  German locale works something strange. In cldr there is a point for after a part of month for example "Mo." but in all cases of Intl there is no a point case and the failed test includes checking against the point case. I think about hardcoding as a solution.
- [x]  `getAmMarker` and `getPmMarker` always returns AM and PM but actually for other locales there are other values too. Intl implementation fixes the problem.
- [x] There are ids for localized strings from `qx.locale` methods which are started with **cldr_** prefix. If I get rid of cldr should I rename this prefix on something else?
- [x] `getTimeFormat` ids are ["short", "medium", "long", "full"] or ["HHmm", "HHmm", "HHmmss", "HHmmsszz"]
- [x] I think more tests are needed.

You can see [here](https://github.com/qooxdoo/qooxdoo/compare/master...goldim:qooxdoo:replace-cldr-by-intl) what I've done so far.
Any suggestions, advices and help will be appreciated.

**Comment by johnspackman** [comment]
> getDateTimeFormat works only for cases which are hardcoded in Cldr.js file of compiler. Other format symbols don't work.

Can we just have a simple mapping (I guess per locale) for these hardcoded strings?

>  I have only one failed test currently with de_DE locale... but in all cases of Intl there is no a point case and the failed test includes checking against the point case. I think about hardcoding as a solution.

Do we need to hardcode a solution for Germany?  EG if the Intl API doesnt handle it, then is it fair to say that most people would not notice the specialism?

> If I get rid of cldr should I rename this prefix on something else?

Sounds reasonable to me - we don't expose CLDR directly, we just use it so IMHO these are internal identifiers

**Reply by goldim** [reply]
> Do we need to hardcode a solution for Germany? EG if the Intl API doesnt handle it, then is it fair to say that most people would not notice the specialism?

I would like to support BC. Yes, I could just use what is Intl returns but it is broken BC. Fortunately we have German developers. They could clarify this moment.

> Sounds reasonable to me - we don't expose CLDR directly, we just use it so IMHO these are internal identifiers

I am afraid if these identifiers are overwritten in .po files and is it possible? Bc there is internal localize flag. AFAIK there are two cases: translation and localization and .po files only for translation? And if I change then what you think about **intl** prefix?

> Can we just have a simple mapping (I guess per locale) for these hardcoded strings?

I was going to do this way for these values (which listed in file Cldr.js) but I wanted to tell that maybe improve this function looks like it doesn't work as supposed.

**Reply by goldim** [reply]
@johnspackman Sorry, I was wrong about `getDateTimeFormat`. It does work. And most difficult part to redo. Removed this task from the list.

**Comment by goldim** [comment]
> I have only one failed test currently with de_DE locale. Intl API has 3 types of weekday: narrow, short, long. Narrow - first letter of weekday, short - usually 2 or 3 letter words, and long - full name. In Cldr there abbreviated, wide and abbreviated. Cldr has another short option but it is not used. I tested with other like french, russian, english and they are fine. German locale works something strange. In cldr there is a point for after a part of month for example "Mo." but in all cases of Intl there is no a point case and the failed test includes checking against the point case. I think about hardcoding as a solution.

I think I found a solution of it.

**Reply by goldim** [reply]
Now I know how to get words with dot in de_DE locale but still have a problem https://github.com/qooxdoo/qooxdoo/issues/10664.

**Comment by goldim** [comment]
Sadly there are no `yQQQ` and `yQ` in Intl API. For `yQ` is only one row fortunately but for `yQQQ` is huge. I think to just include list of values for it.

**Comment by hkollmann** [comment]
it's avaiable in branch_8_0-beta
